### PR TITLE
Prepare release 5.12.0

### PR DESCRIPTION
**Pull request by johnwiseheart**
#### Overview of change:

Changes required to release 5.12.0.

Steps for me following this:

* Create a release on github
* Publish the release locally
* Fix docs

#### Is there anything you'd like reviewers to focus on?

This should just work right? I haven't specifically smoke tested it - thoughts? Suppose you could install and build it locally, then force it to use your local version through `yarn link` magic to test? Our tests in repo are pretty thorough, but not sure they will cover non-rule bugs added.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,150 @@
...
 
+## v5.12.0
+
+## :tada: Features
+
+-   [feature] Configure format from file (#4155)
+-   [feature] Make cli expect array for rules dir (#3788)
+-   [feature] Set js rules to all valid active rules (#3641)
+-   [improvement] Add tests for typescript@2.9 and typescript@3.0 (#4053)
+-   [new-fixer][type-literal-delimiter](https://palantir.github.io/tslint/rules/type-literal-delimiter/) (#3964)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: Nit: formatting is off for type-literal-delimiter (missing backticks). Here and lower down too.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,150 @@
...
+-   [bugfix] Fix configuration tests when comparing rules (#4346)
+-   [bugfix] Fix semicolon: Cannot read property 'initializer' of undefined (#4351)
+-   [bugfix] Fix whitespace rule in importClause with 'as' statement (#4249)
+-   [bugfix][`no-invalid-this`](https://palantir.github.io/tslint/rules/no-invalid-this/)](https://palantir.github.io/tslint/rules/no-invalid-this/) rule fixes false positives on method-like syntax and false negatives on nested functions (#4034)
+-   [bugfix][`one-line`](https://palantir.github.io/tslint/rules/one-line/): Check block in a case clause (#3619)
+-   [bugfix][`whitespace`](https://palantir.github.io/tslint/rules/whitespace/) Don't require space between import and paren with in type imports (#4243)
+-   [bugfix] remove [`no-unused-variable`](https://palantir.github.io/tslint/rules/no-unused-variable/) from `tslint:all` (#4079)
+-   [bugfix][`callable-types`](https://palantir.github.io/tslint/rules/callable-types/)](https://palantir.github.io/tslint/rules/callable-types/) support interfaces containing a single construct signature (#4291) (#4352)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: > `..-types/)](https://...`

Formatting typo: extra `]` in the mix

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,150 @@
...
+-   [new-rule-option] `{ singleLine: "always" | "never"}` for rule [type-literal-delimiter](https://palantir.github.io/tslint/rules/type-literal-delimiter/) (#3964)
+-   [new-rule-option]: `"snake-case"` in `"file-name-casing"` (#4081)
+-   [new-rule] Enforce Blank Line after File Header (#3740)
+-   [new-rule][`ban-ts-ignore`](https://palantir.github.io/tslint/rules/ban-ts-ignore/) (#3728)
+-   [new-rule][`function-constructor`](https://palantir.github.io/tslint/rules/function-constructor/) (#4198)
+-   [new-rule][`increment-decrement`](https://palantir.github.io/tslint/rules/increment-decrement/) (#3575)
+-   [new-rule][`no-default-import`](https://palantir.github.io/tslint/rules/no-default-import/) (#4023)
+-   [new-rule] `no-unnecessary-constructor` (#3647)
+-   [new-rule] comment-type (#4008)
+-   [new-rule] unnecessary-bind (#3588)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: Nit: these three are missing links, and two are missing backticks

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,150 @@
...
+
+## :hammer_and_wrench: Bugfixes & enhancements
+
+-   [api] Allow `exclude` to work with node API (#4143)
+-   [api] Upgrade TSLint to support TS 3.1 (#4274)
+-   [bugfix](#4192)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: Nit: odd quirk of the markdown flavor, this isn't showing a link

image

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,150 @@
...
+-   [bugfix] Added exception for generators in [`return-undefined`](https://palantir.github.io/tslint/rules/return-undefined/) rule (#4370)
+-   [bugfix] Allow unknown in 'no-object-literal-type-assertion' rule. (#4362)
+-   [bugfix] Completed docs rule can check single-line comments (#3841)
+-   [bugfix] Correct the auto-fix code of [`prefer-method-signature`](https://palantir.github.io/tslint/rules/prefer-method-signature/) rule, now capable to handle (#4066)
+-   [bugfix] Default logger.error should log to stderr. (#3982)
+-   [bugfix] Files an issue where TSLint doesn't correctly resolve packages in a `node_modules` folder that are symlinked to another location where their dependencies are satisfied. (#4295)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: > Files

Typo: Fixes?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,150 @@
...
+-   [bugfix] Fix configuration tests when comparing rules (#4346)
+-   [bugfix] Fix semicolon: Cannot read property 'initializer' of undefined (#4351)
+-   [bugfix] Fix whitespace rule in importClause with 'as' statement (#4249)
+-   [bugfix][`no-invalid-this`](https://palantir.github.io/tslint/rules/no-invalid-this/)](https://palantir.github.io/tslint/rules/no-invalid-this/) rule fixes false positives on method-like syntax and false negatives on nested functions (#4034)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: > `..-this/)](https://...`

Formatting typo: extra `]` in the mix

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,150 @@
...
+-   [bugfix][`one-line`](https://palantir.github.io/tslint/rules/one-line/): Check block in a case clause (#3619)
+-   [bugfix][`whitespace`](https://palantir.github.io/tslint/rules/whitespace/) Don't require space between import and paren with in type imports (#4243)
+-   [bugfix] remove [`no-unused-variable`](https://palantir.github.io/tslint/rules/no-unused-variable/) from `tslint:all` (#4079)
+-   [bugfix][`callable-types`](https://palantir.github.io/tslint/rules/callable-types/)](https://palantir.github.io/tslint/rules/callable-types/) support interfaces containing a single construct signature (#4291) (#4352)
+-   [bugfix][`import-sources-order`] Setting `import-sources-order: any` no longer invalidates `grouped-imports: true` (#4374)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: > [`import-sources-order`]

Is this missing a link?

**Comment by johnwiseheart** [comment]
Oops a lot of these were generated, going to take a closer look at the list and fix up your catches.

**Comment by johnwiseheart** [comment]
@JoshuaKGoldberg got them all, I think. This should be good to approve now, but I probably won't have time to run the post-merge steps until tomorrow morning (london time), so lets wait until then to merge.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,150 @@
...
+-   [new-rule-option][`import-blacklist`](https://palantir.github.io/tslint/rules/import-blacklist/): support blacklisting specific named exports (#3926)
+-   [new-rule-option][`type-literal-delimiter`](https://palantir.github.io/tslint/rules/type-literal-delimiter/): Add `{ singleLine: "always" | "never"}` (#3964)
+-   [new-rule-option][`file-name-casing`](https://palantir.github.io/tslint/rules/file-name-casing/): Add `snake-case` option (#4081)
+-   [new-rule][`file-header`](https://palantir.github.io/tslint/rules/file-header/) (#3740)
+-   [new-rule][`ban-ts-ignore`](https://palantir.github.io/tslint/rules/ban-ts-ignore/) (#3728)
+-   [new-rule][`function-constructor`](https://palantir.github.io/tslint/rules/function-constructor/) (#4198)
+-   [new-rule][`increment-decrement`](https://palantir.github.io/tslint/rules/increment-decrement/) (#3575)
+-   [new-rule][`no-default-import`](https://palantir.github.io/tslint/rules/no-default-import/) (#4023)
+-   [new-rule][`no-unnecessary-constructor`](https://palantir.github.io/tslint/rules/no-unnecessary-constructor/) (#3647)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: Sorry, I renamed it to `unnecessary-constructor` during the PR and forgot to update the changelog entry ☹

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,150 @@
...
+-   [new-fixer][`type-literal-delimiter`](https://palantir.github.io/tslint/rules/type-literal-delimiter/) (#3964)
+-   [new-rule-option] Add whitespace rule for open braces
+-   [new-rule-option][`file-name-casing`](https://palantir.github.io/tslint/rules/file-name-casing/): Add file name whitelist (#4206)
+-   [new-rule-option][`object-literal-sort-keys`](https://palantir.github.io/tslint/rules/object-literal-sort-keys/): `locale-compare` (#4193)
+-   [new-rule-option][`strict-boolean-expressions`](https://palantir.github.io/tslint/rules/strict-boolean-expressions/) adds `allow-enum` to allow using enums in boolean expressions (#3604)
+-   [new-rule-option][`import-blacklist`](https://palantir.github.io/tslint/rules/import-blacklist/): support blacklisting specific named exports (#3926)
+-   [new-rule-option][`type-literal-delimiter`](https://palantir.github.io/tslint/rules/type-literal-delimiter/): Add `{ singleLine: "always" | "never"}` (#3964)
+-   [new-rule-option][`file-name-casing`](https://palantir.github.io/tslint/rules/file-name-casing/): Add `snake-case` option (#4081)
+-   [new-rule][`file-header`](https://palantir.github.io/tslint/rules/file-header/) (#3740)
```
Comment (JoshuaKGoldberg) [inline CHANGELOG.md]: > file-header

This rule already exists; the PR was the enhancement "Option for file-header to enforce a blank line after the header"

**Pull request merged**